### ISSUE TITLE
Wrong gradients in a Haiku network

### ISSUE THREAD

**Issue by SNMS95** [google-deepmind/dm-haiku#741]
Hey guys, I was trying to use haiku to create a convolutional neural network architecture (to reproduce a paper, whose implementation was in TF2.0). The CNN works correctly, however, when I use `jax.test_util.check_grads`, there seems to be an error. The code is as follows:

```
import jax.numpy as np
import jax
from functools import partial


class CNNParameterization(hk.Module):

    def __init__(self):
        super().__init__()
        self.layers = self._build_layers()
        
    def _build_layers(self):
        activation = jax.nn.leaky_relu
        Nx = 64
        Ny = 64
        total_resize = onp.prod((1, 2, 2, 2, 1 ))
        h = Nx // total_resize
        w = Ny // total_resize

        # global normalization
        def glob_norm(x):
            mean = np.mean(x)
            variance = np.var(x)
            return (x - mean) / np.sqrt(variance + 1e-6)

        def upsample(x, resize):
            new_shape = (1, resize*x.shape[1], resize*x.shape[2], x.shape[3])
            x = jax.image.resize(x, new_shape, method='bilinear')
            return x
        offset_scale_factor = 1.0

        def add_offset_fn(x):
            offset_weights = hk.get_parameter(
                            name="offset",
                            shape=x.shape, init=hk.initializers.Constant(0.0))
            return x + offset_weights * offset_scale_factor

        # Convert to Module to use it multiple times by creating instances
        AddOffset_obj = hk.to_module(add_offset_fn)

        layers = []
        self.latent_params = hk.get_parameter(
                        "beta", shape=(128, ),
                        init=hk.initializers.RandomNormal())
        dense_output_size = 32*w*h
        dense_init = hk.initializers.Orthogonal(
                        scale=1.0*onp.sqrt(onp.max(
                                (dense_output_size/128, 1)
                            )))
        dense_layer = hk.Linear(output_size=dense_output_size,
                                name='dense_layer',
                                w_init=dense_init)
        layers.append(dense_layer)
        # Reshape preserves batch dimension
        layers.append(hk.Reshape((h, w, 32),
                                 name="reshape"))
        counter = 0
        for resize, conv_filters in zip((1, 2, 2, 2, 1), \
                                        (32, 16, 8, 4, 1)):
            layers.append(activation)
            upsample_fn_instance = partial(upsample, resize=resize)
            layers.append(upsample_fn_instance)
            layers.append(glob_norm)
            layers.append(hk.Conv2D(output_channels=conv_filters,
                                    kernel_shape=(5, 5),
                                    padding='SAME',
                                    name='conv_layer',
                                    w_init=hk.initializers.VarianceScaling()))
            layers.append(AddOffset_obj(name=f'add_offset_{counter}'))
            counter += 1
        return layers

    def __call__(self, model_input: jax.Array = None):
        """Forward pass.

        The model input is unused.
        """
        del model_input

        x = self.latent_params
        for layer_no, layer in enumerate(self.layers):
            if layer_no == 1:  # Only for reshaping layer
                x = layer(x.reshape((1, ) + layer.output_shape))
            else:
                x = layer(x)
        x = x.ravel()
        return x

# Test the gradients
def mapping_fn(x):
     result = CNNParameterization()(x)
     return result

model_input = np.ones((100, 3))
forward_pass_pure = hk.without_apply_rng(
    hk.transform_with_state(mapping_fn))
init_params, init_state = forward_pass_pure.init(x=model_input,
                                                 rng=rng_key)
forward_func = jax.jit(forward_pass_pure.apply)

def dummy_func(params, state, x):
    return forward_func(params, state, x)[0].mean()

check_grads(dummy_func, (init_params, init_state, model_input),
            order=2, eps=1e-4)
```
